9.0 cresus import

open-net-sarl opened the pull request:
I basically reenabled the module on 9.0 and fixed what came up. And added tests. I also had to show `BaseModel.load` the door because it matches account codes too broadly (proof: [Travis results when you revert that change](https://travis-ci.org/open-net-sarl/l10n-switzerland/builds/144994088)). While at it, I improved the UI, just like [this comment](https://github.com/OCA/l10n-switzerland/pull/156#issuecomment-135695263) suggested on a sister module (which I'll be done porting too soon). More significantly: I  made it so that the module could import more complicated moves, which Crésus indicates with an ellipsis in the account code columns, which also prompted me to group lines into moves based on the “pce” number rather than the date.

For the record: I had data coming from Crésus Comptabilité 3.3m PRO to test around, and that's why I came across ellipses. I know the module was designed for Crésus Salaires. Anyway, it's probably better that way. I didn't encounter lines with negative “amount”, but it was already supported in the 8.0 module, so I kept it.

vrenaville committed:
[FIX] Do not set a default partner on move line

Conflicts:
	l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py

Commit:
[FIX] Use xpath for `...position=*>` in xml

Commit:
reenable module

Commit:
remove period_id

9.0 doesn't have it

Commit:
add an input date format

It seems the sample data I have received uses it

Commit:
adapt to 9.0 database structure

Commit:
add support for input multi-line moves

They appeared im my sample data.

Commit:
disable logging

Commit:
update readme

Commit:
update and beautify embedded html help

Commit:
update __openerp__.py

Commit:
make self.state a selection

not strictly necessary but nicer that way

Commit:
improve UI

Commit:
make into a popup

maybe just disabling new/discard would be enough, but I found no way to do it
simply

Commit:
remove "view imported moves" and short-circuit

Commit:
pep8

Commit:
lint

Commit:
switch to recommended module layout

Commit:
update translations

Commit:
add unit test

Commit:
fix condition for previous_pce

'' is False but my data uses it

Commit:
rewrite to use create instead of load

The thing is: there's no way to control how load does the job of matching strings to accounts, and it seems it matches '1020' with '10203' because it starts with '1020', which constitutes a serious bug. Out goes load.

Commit:
add the error popup back

Commit:
lint

Commit:
add a test case for negative amount

Commit:
update translation

Commit:
add case for yesterday's bug

See 34626658d56920dd83b83d0cecfc55bba8f9d659

yvaucher commented on l10n_ch_import_cresus/__openerp__.py:
```diff
@@ -1,36 +1,24 @@
+    'summary': 'Allows to import Cresus Salaires .txt files containing \
+                journal entries into Odoo.',
+    'version': '9.0.1.0.0',
+    'category': 'uncategorized',
```
What about "Human Resources" as category ?

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:
```diff
@@ -137,194 +71,150 @@ def _parse_csv(self):
+                continue
+        else:
+            raise exceptions.ValidationError(
+                "Can't parse date '%s'" % date_string)
```
The error message should be translatable.

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:
```diff
@@ -137,194 +71,150 @@ def _parse_csv(self):
+        account_obj = self.env['account.account']
+        account = account_obj.search([('code', '=', code)], limit=1)
+        if not account:
+            raise exceptions.MissingError("No account with code %s" % code)
```
Error message to wrap with `_( ... )`

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:
```diff
@@ -137,194 +71,150 @@ def _parse_csv(self):
+    def _find_tax(self, typtvat, account):
+        tax_obj = self.env['account.tax']
+        if not typtvat or account.user_type_id.include_initial_balance:
+            return tax_obj.browse()
```
no need to call the `browse()`

`tax_obj` is already a record set.

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:
```diff
@@ -137,194 +71,150 @@ def _parse_csv(self):
+CSV file is malformed
+Please choose the correct separator
+the error detail is:
+%r''' % error)
```
Why breaking the indent ? The format:

```
    (" ... "
     " ...")
```

is  valid.

lbeopennet replied:
does it also need `_( .... )` ?

yvaucher replied:
Indeed it does for all Odoo exceptions.

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:
```diff
@@ -1,131 +1,65 @@
-        :param messages: return of BaseModel.load messages key
-        :returns: formatted string
+    ODOO_LINE_ARGS = {'account_id', 'partner_id', 'name',
+                      'tax_line_id', 'analytic_account_id'}
```
This isn't used, to remove ?

yvaucher commented on l10n_ch_import_cresus/i18n/fr.po:
```diff
@@ -1,163 +1,202 @@
-#: field:account.cresus.import,help_html:0
+#: model:ir.ui.view,arch_db:l10n_ch_import_cresus.cresus_move_line_importer_form
+msgid "Import file"
+msgstr "Import file"
```
Keep the string empty when it's not translated.

yvaucher commented on l10n_ch_import_cresus/tests/test_import.py:
```diff
@@ -0,0 +1,78 @@
+        user_type = self.env['account.account.type'].search(
+            [('include_initial_balance', '=', False)], limit=1)
+        for code in '1000 1010 1210 2200 2800 2915 6512 6513 6642 '\
+                    '9100 10101'.split():
```
Why using a string instead of a list ?

lbeopennet replied:
tbh the main reason is that I was too lazy to add a comma manually. It does the same in less space tho, and it echoes Ruby's `%w,1000 1010 1210 2200,` quite nicely IMHO

yvaucher replied:
Here you go to save your fingers:

```
['1000', '1010', '1210', '2200', '2800', '2915', '6512', '6513', '6642', '9100', '10101']
```

yvaucher commented on l10n_ch_import_cresus/tests/test_import.py:
```diff
@@ -0,0 +1,78 @@
+''')})
+        wizard.import_file()
+
+        res = move_obj.browse(wizard.imported_move_ids.ids)
```
`res = move_obj.browse(wizard.imported_move_ids.ids)` => `res =  wizard.imported_move_ids`

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:
```diff
@@ -1,131 +1,65 @@
-            )
-        return "\n \n".join(res)
+    @staticmethod
+    def make_line(debit_amount, credit_amount, **kwargs):
```
`prepare_line` would be more suited

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:
```diff
@@ -137,194 +71,150 @@ def _parse_csv(self):
+            previous_date = line_cresus['date']
+
+            recto_amount = float(line_cresus['amount'].replace('\'', '')
+                                                      .replace(' ', ''))
```
Maybe you could use

```
import locale

locale.setlocale(locale.LC_ALL, 'fr_CH')
locale.atof(line_cresus['amount'])
```

lbeopennet replied:
fr_CH doesn't consider the space as a valid thousands separator so it won't do the same, and I essentially got it this way from 8.0. I'd rather keep it in this less abstracted way because I'm not sure exactly what Cresus does.

yvaucher replied:
In fact it is a bad idea to use setlocale as it could change it for all threads

But you should use the right tool.

Thus it would be better to use pybabel (thanks @gurneyalex for the tip)

``` python
from babel.numbers import parse_decimale

float(parse_decimal("       1'099.98      ", locale='de_CH'))
```

lbeopennet replied:
Okay that'll do

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:
```diff
@@ -137,194 +71,150 @@ def _parse_csv(self):
+                    partner_id=False,
+                    name=line_cresus['ref'],
+                    tax_line_id=tax.id,
+                    analytic_account_id=analytic_account.id))
```
don't you want to pass the account to the method "make_line" so you can define all those arguments in it ?

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:
```diff
@@ -137,194 +71,150 @@ def _parse_csv(self):
-        return self._load_data(new_data)
+        try:
+            self._import_file()
+        except Exception as exc:
```
Could you provide a more precise type of exception to catch ?

lbeopennet replied:
The point is to catch any unexpected error and show it in our interface rather that the default "Odoo Server Error" popup. Got it this way from 8.0.

yvaucher commented on l10n_ch_import_cresus/wizard/l10n_ch_import_cresus_view.xml:
```diff
@@ -1,48 +1,43 @@
+        <field name="model">account.cresus.import</field>
+        <field name="arch" type="xml">
+            <form string="Move / Move lines importer">
+                <footer>
```
Could you move this footer at the bottom would make more sense ?

yvaucher replied:
Don't we want also a cancel button ?

lbeopennet replied:
Either nothing has been imported and I can just close the popup (It's all or nothing because of `cr.rollback`) or the import has succeeded and I'm automatically sent to a view on the moves I just imported, and I get a chance to manually delete some or all of them. The 'automatically sent' part was my idea tho, there was a IMHO-redundant 'view imported moves' button before so i can revert that if needs be.

yvaucher commented:
@lbeopennet Hello thanks for having added some tests.

Please see my comments

Commit:
let category = 'accounting'

That's what I would rather classify it as.

'hr' doesn't make a lot of sense because the menus are in accounting and
reserved for adviser, and what it really imports is accounting lines. There was
a bit of confusion as the original need for the module is to import
salary-related lines from Crésus Salaires (which are more frequently fed to
Crésus Comptabilité but it's not meant to be exclusive) in order to pay those
salaries, but the module can also be used to import lines from Crésus
Comptabilité in order to completely migrate, which has nothing to do with HR.

Commit:
don't break indent

Commit:
I give up on the translation

Someone who understands the tool please do it

Commit:
unnecessary calls to browse

Commit:
put <footer> at bottom of file

Commit:
rename helpers

Commit:
update translatable fields

Commit:
shouldn't have commented out my asserts

Commit:
DRY prepare_{move,line}

Commit:
pep8

Commit:
extra pep8

yvaucher commented:
About translations, you should probably exports all translations (after having added all `_( ... )`  to error messages) in a template (Developper mode / Settings / Translations / Export .
Use new language and select this module. This should create you a `.pot` file.

Thus use a po editor. For example poedit (free opensource) and open the `.fr` file and update it with the `.pot` file you just created. Thus you can edit in the GUI. Otherwise you can still use gettext to do the update in commandline.

Commit:
use babel for reading numbers

ghost commented:
About translations, been here done that, but when I do that, even on a test database, I get back a file that stats with this ghost translation:
`#. module: l10n_ch_import_cresus
#: code:addons/l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py:64
#, python-format
msgid "\n"
"                 In order to import your 'Cresus Salaires' .txt                  file you must complete the following requirements : </br>\n"
"                * The accounts, analytical accounts used in the Cresus                 file must be previously created into Odoo  </br>\n"
"                * The date of the entry will determine the period used                 in Odoo, so please ensure the period is created already. </br>\n"
"                * If the Cresus file uses VAT codes (i.e: IPI),                 please make sure you have indicated this code in the                 related Odoo tax (new field).                 Warning, the Odoo tax must be 'tax included'.                 If the tax does not exist you have to create it. </br>\n"
"                * All PL accounts must have a deferral method = 'none'                (meaning: no balance brought forward in the new fiscal year)                and all\n"
"                 Balance sheet account must have a deferral                 method = 'balance'. </br>"
msgstr ""`
which is just not [there](https://github.com/open-net-sarl/l10n-switzerland/blob/9.0-cresus-import/l10n_ch_import_cresus/wizard/l10n_ch_import_cresus.py#L64) ! what's wrong? (I know you won't necessarily guess, maybe it's on our server, and I have to bounce anyway, I'll try on the runbot tomorrow and then come back to complain)

Commit:
use literal list instead of str.split()

Commit:
try translating

Commit:
lint+pep8

ghost commented:
(still testing, hold on)

Commit:
depend on account_accountant

otherwise the menu entry doesn't show up because it is under “Adviser”

Commit:
add more translating _

Commit:
beautify unit test

Commit:
update translation

Commit:
remove redundant translations

seems like they get overwritten when I reexport, and it's generic
terminology anyway

Commit:
change popup title

Same as the button, seems logical.

Commit:
remove unused form string

Commit:
update french translation

ghost commented:
ok, so I got around all of that, the runbot was immeasurably helpful. Hey @yvaucher, mind reviewing me again?

Commit:
fix problem with exceptions during unit tests

calling the inner function so the self.cr.rollback() part doesn't kick in,
as the AccountingCresusImport itself didn't when this cursor was created.
That's why I created _import_file originally, I must've forgotten to
replace the call then.

yvaucher commented on l10n_ch_import_cresus/models/account_tax.py:
```diff
@@ -0,0 +1,13 @@
+    _inherit = "account.tax"
+
+    tax_cresus_mapping = fields.Char(
+        size=128,
```
Just asking, do you know if the size 128 is a limitation of cresus ? if not size attribute can be dropped.

lbeopennet replied:
I have no idea. Besides, maybe it's only the case on earlier versions, so better drop it.

yvaucher replied:
@lbeopennet you are right, I just checked with @vrenaville and it can be dropped. Size was mandatory in older version of Odoo and there is no reason to keep it.

yvaucher commented on l10n_ch_import_cresus/tests/cresus.txt:
```diff
@@ -0,0 +1,12 @@
+12.01.02	1010	6642	3	Frais de déplacement - Tolochenaz	-15.00
+14.01.02	6512	...	4	Salt, (TVA) net, TVA = 13.17	173.38
+14.01.02	2200	...	4	Salt, 7.6% de TVA (TVA)	13.17	VAT
+14.01.02	...	1010	4	Salt Total, (TVA)	186.55
```
Thanks that's cleaner to have an exemple file on it's own :+1:

lbeopennet replied:
Thanks ^^

yvaucher commented:
@lbeopennet Thanks for all the changes LGTM now

:+1:

Commit:
Remove legacy size limit

leemannd commented:
:+1:
